bookkeeper-server/src/main/java/org/apache/bookkeeper/tools/cli/commands/bookie/SanityTestCommand.java

```diff
@@ -81,59 +86,126 @@ public boolean apply(ServerConfiguration conf, SanityFlags cmdFlags) {
+        } catch (BKException | IOException | InterruptedException e) {
+            LOG.warn("Failed to initialize bookkeeper client", e);
+            result.completeExceptionally(e);
+            return result;
+        }
 
+        bk.asyncCreateLedger(1, 1, BookKeeper.DigestType.MAC, new byte[0], (rc, lh, ctx) -> {
+            if (rc != BKException.Code.OK) {
+                LOG.warn("ledger creation failed for sanity command {}", rc);
+                result.completeExceptionally(BKException.create(rc));
+                return;
+            }
+            List<CompletableFuture<Void>> entriesFutures = new ArrayList<>();
             for (int i = 0; i < cmdFlags.entries; i++) {
                 String content = "entry-" + i;
-                lh.addEntry(content.getBytes(UTF_8));
-            }
-
-            LOG.info("Written {} entries in ledger {}", cmdFlags.entries, lh.getId());
-
-            // Reopen the ledger and read entries
-            lh = bk.openLedger(lh.getId(), BookKeeper.DigestType.MAC, new byte[0]);
-            if (lh.getLastAddConfirmed() != (cmdFlags.entries - 1)) {
-                throw new Exception("Invalid last entry found on ledger. expecting: " + (cmdFlags.entries - 1)
-                                        + " -- found: " + lh.getLastAddConfirmed());
+                CompletableFuture<Void> entryFuture = new CompletableFuture<>();
+                entriesFutures.add(entryFuture);
+                lh.asyncAddEntry(content.getBytes(UTF_8), (arc, alh, entryId, actx) -> {
+                    if (arc != BKException.Code.OK) {
+                        LOG.warn("ledger add entry failed for {}-{}", alh.getId(), arc);
+                        entryFuture.completeExceptionally(BKException.create(arc));
+                        return;
+                    }
+                    entryFuture.complete(null);
+                }, null);
             }
+            CompletableFuture<LedgerHandle> lhFuture = new CompletableFuture<>();
+            CompletableFuture<Void> readEntryFuture = new CompletableFuture<>();
+
+            FutureUtils.collect(entriesFutures).thenCompose(_r -> {
+                bk.asyncOpenLedger(lh.getId(), BookKeeper.DigestType.MAC, new byte[0], (orc, olh, octx) -> {
+                    if (orc != BKException.Code.OK) {
+                        LOG.warn("open sanity ledger failed for {}-{}", lh.getId(), orc);
+                        lhFuture.completeExceptionally(BKException.create(orc));
+                        return;
+                    }
+                    long lac = olh.getLastAddConfirmed();
+                    if (lac != (cmdFlags.entries - 1)) {
+                        lhFuture.completeExceptionally(new Exception("Invalid last entry found on ledger. expecting: "
+                                + (cmdFlags.entries - 1) + " -- found: " + lac));
+                        return;
+                    }
+                    lhFuture.complete(lh);
+                }, null);
```

Comment:
We'd better add the `customMetadata` for the created ledger in order to find out orphan ledgers if the ledger delete failed.

Reply:
this is existing code where we don't have handling for failed delete. this change only changes blocking IO to async to avoid thread-blocking while checking sanity. 
So, if we need any enhancement into existing codebase then we can address it in separate PR.

Reply:
This PR makes the case worse. The existing code only triggers the sanity test by the bookkeeper shell, but from this PR on, the bookie will trigger sanity test every 60s

Reply:
well running sanity is executing one operation of what bookie suppose to do with high throughput : creating ledger/add entries/read entries/delete ledger. so, one operation per minute should be fine. the only thing is we should not have any blocking IO and that's the reason, I have changed sanityTestCommand implementation with async behavior.

Reply:
> I'm not sure whether we need a flag to turn on/off the sanity test

sure, we can add the flag and keep it disable by default and user can enable based on the need.

Reply:
added the flag `sanityCheckMetricsEnabled` and disabled the behavior by default.